# Введение в регрессионный анализ: интенсив по Python

*Алла Тамбовцева*

## Семинар 1. Работа с массивами NumPy: часть 1

### Немного теории: функции, методы и атрибуты

Создадим объекты трех разных типов:

* в переменную `grades` сохраним **список** из целых чисел (тип `list`)
* в переменную `rating` сохраним **массив** из дробных чисел (тип `numpy array`)
* в переменную `lullaby` сохраним **строку** с текстом (тип `string`)

In [1]:
import numpy as np

In [2]:
grades = [10, 9, 6, 5, 8]
rating = np.array([4.5, 9.8, 6.5, 8.2, 9.3, 4.8, 5.2, 6.3])
lullaby = "Спят усталые питоны, змеи спят... И усталые студенты спать хотят"

> Квадратные скобки вокруг элементов в `grades` гарантируют создание именно списка. В Python каждый тип скобок соответствует определенному типу, например, если перечислить числа в круглых скобках, получится кортеж (тип `tuple`), а если в фигурных – получится множество (тип `set`). При создании строки с текстом роль скобок выполняют кавычки – они отмечают, где текст начинается и заканчивается. А в случае с массивом одних скобок недостаточно. Чтобы получить массив из библиотеки NumPy, эту библиотеку нужно сначала импортировать (для удобства импортировали с сокращенным названием `np` выше), а затем задействовать функцию `array()` оттуда.

In [3]:
print(grades)
print(rating)
print(lullaby)

[10, 9, 6, 5, 8]
[4.5 9.8 6.5 8.2 9.3 4.8 5.2 6.3]
Спят усталые питоны, змеи спят... И усталые студенты спать хотят


> Когда массив выводится на экран через `print()`, в нем убираются запятые между элементами, это нормально. Массив не перестает быть массивом, это позволяет уменьшить число символов на экране и сделать выдачу более читаемой. При выводе списков такого не происходит, для более простых структур дополнительное «оформление» не предусмотрено.

На **функцию** в Python можно смотреть как на команду, которая выполняет определенную задачу и работает универсальным образом на разных типах объектов. Объект, к которому мы применяем функцию, называется **аргументом** функции и указывается в круглых скобках.

Например, функция `len()` возвращает длину любой последовательности – считает количество элементов в ней:

In [4]:
print(len(grades)) # 5 чисел
print(len(rating)) # 8 чисел
print(len(lullaby)) # 64 символа

5
8
64


Функция `sorted()` возвращает список с отсортированными элементами последовательности:

In [5]:
print(sorted(grades)) # отсортированные элементы списка
print(sorted(rating)) # отсортированные элементы массива
print(sorted(lullaby)) # отсортированные символы в тексте

[5, 6, 8, 9, 10]
[np.float64(4.5), np.float64(4.8), np.float64(5.2), np.float64(6.3), np.float64(6.5), np.float64(8.2), np.float64(9.3), np.float64(9.8)]
[' ', ' ', ' ', ' ', ' ', ' ', ' ', ' ', ' ', ',', '.', '.', '.', 'И', 'С', 'а', 'а', 'а', 'д', 'е', 'е', 'е', 'е', 'з', 'и', 'и', 'л', 'л', 'м', 'н', 'н', 'о', 'о', 'п', 'п', 'п', 'п', 'с', 'с', 'с', 'с', 'с', 'т', 'т', 'т', 'т', 'т', 'т', 'т', 'т', 'т', 'т', 'у', 'у', 'у', 'х', 'ы', 'ы', 'ы', 'ы', 'ь', 'я', 'я', 'я']


Функция `sum()` возвращает сумму элементов последовательности:

In [6]:
print(sum(grades))
print(sum(rating))

38
54.599999999999994


> Вообще сумма элементов `rating` равна 54.6, но Python выводит такой результат из-за *проблемы представления*, возникающей при работе с десятичными дробями. Python переводит числа из нашей привычной десятичной системы в двоичную, выполняет вычисления, затем результат снова переводит в десятичную. В ходе этих манипуляций накапливается погрешность. Мы пока с ней смиримся, а вообще есть модули, позволяющие выставить интересующее нас число знаков после точки и избежать этой проблемы.

**Метод** – функция на объекте определенного типа. Так, у списков будет один набор методов (добавление, удаление, изменение элементов), у строк – другой (обработка текста), у числовых массивов  – третий (манипуляции с числами). 

В отличие от функций метод пишется не перед названием объекта, к которому применяется, а **после** – чтобы Python, читающий код слева направо, сначала понимал, с каким типом объекта он имеет дело.

In [7]:
# методы на строках

# .upper() – возвращает копию строки, где все буквы заглавные
# .lower() – возвращает копию строки, где все буквы строчные
# .split() – разбивает строку на список строк по заданному разделителю,
# в круглых скобках можно указать разделитель

print(lullaby.upper())
print(lullaby.lower())
print(lullaby.split("..."))

СПЯТ УСТАЛЫЕ ПИТОНЫ, ЗМЕИ СПЯТ... И УСТАЛЫЕ СТУДЕНТЫ СПАТЬ ХОТЯТ
спят усталые питоны, змеи спят... и усталые студенты спать хотят
['Спят усталые питоны, змеи спят', ' И усталые студенты спать хотят']


In [8]:
# методы на массивах

# .mean() – вычисляет среднее
# .var() – вычисляет выборочную дисперсию
# .reshape() – изменяет форму массива, 
# разбивая его более мелкие массивы с заданным числом элементов

print(rating.mean())
print(rating.var())
print(rating.reshape((2, 4))) # 2 массива по 4 элемента

6.825
3.6743750000000004
[[4.5 9.8 6.5 8.2]
 [9.3 4.8 5.2 6.3]]


**Важно.** Несмотря на то, что в круглых скобках после названия метода часто ничего не указывается (если метод не предполагает аргументов), сами скобки убирать нельзя. Их наличие обеспечивает «запуск» метода, то есть его применение к объекту. Если их стереть, Python просто сообщит нам о том, что метод существует, не выполняя заданную операцию.

In [9]:
# .mean – для объекта rating типа numpy.ndarray, который
# сохранен в ячейке памяти 0x104feb870, определен метод .mean

print(rating.mean)

# .mean() – метод применяется к rating и выдает результат

print(rating.mean())

<built-in method mean of numpy.ndarray object at 0x1077af750>
6.825


**Атрибут** – фиксированная характеристика или свойство объекта. Он также вызывается через точку после названия объекта, однако не требует круглых скобок. У базовых объектов типа списков или строк атрибутов нет, а у массивов есть:

In [10]:
# атрибуты массивов
# .dtype – тип элементов (data type)
# .size – число элементов
# .ndim – число измерений (n dimensions)

print(rating.dtype)
print(rating.size)
print(rating.ndim)

float64
8
1


Как ранее было отмечено, круглые скобки обеспечивают выполнение действия (`.mean()` – вычисление среднего, `.reshape()` – изменение структуры массива), а здесь действия нет, мы просто забираем какую-то готовую сохраненную ранее характеристику массива, поэтому и скобок нет.

**Очень дополнительно – для желающих.** Более наглядный пример. Мы захотели создать свой новый тип данных – «студент» (формально – объект класса, который мы назовем `Student`). Закрепим за объектом этого типа атрибуты – обязательные характеристики студента – фамилию `.surname`, имя `.name`, номер курса `.grade`. Далее опишем методы – функции, которые будут работать только на объектах класса «студент». Например, метод `.info()`, который будет выводить строчку с фамилией, именем и курсом студента, метод `.upgrade()`, который будет переводить студента на следующий курс – увеличивать номер курса на 1, и метод `.expel()`, который будет «отчислять» студента – заменять номер курса на 0.

In [11]:
# часть 1: описываем класс Student
# тут без подробных пояснений – можно после пар поспрашивать,
# важен результат в части 2

class Student:
    def __init__(self, surname, name, grade):
        self.surname = surname
        self.name = name
        self.grade = grade
    def info(self):
        print(f"{self.surname} {self.name} {self.grade} курс")
    def upgrade(self):
        self.grade = self.grade + 1
    def expel(self):
        self.grade = 0

Теперь мы можем создавать объекты класса `Student` и применять к нему методы, которые написали (точно так же ранее программисты написали класс `ndarray` в рамках библиотеки `numpy` и определили на нем много полезных методов).

In [12]:
# создаем запись для Пети на 1 курсе
# через метод .info() просим его описать

petya = Student("Змеевкин", "Петя", 1)
petya.info()

Змеевкин Петя 1 курс


In [13]:
# через метод .upgrade() переводим Петю на 2 курс
# запрашиваем атрибуты и видим, что курс изменен

petya.upgrade()
print(petya.surname)
print(petya.name)
print(petya.grade)

# кто не гуманист – можете отчислить Петю
# через метод .expel()

Змеевкин
Петя
2


### Задача 1. Победит сильнейший

Создайте следующие массивы:

* массив `numbers` из следующих целых чисел: 3, -1, 7, 8, 10, 11, 2, 7, 5
* массив `scores` из следующих чисел: 10, 8.5, 7.8, 9, 9.2, 6.7
* массив `names` из следующих строк: Алёна, Борис, Василий, Георгий, Даниил
* массив `column` из следующих значений: 1, 2, 3, 4, 5, id6

Выведите массивы и типы их элементов на экран. Подумайте, почему задача названа таким образом.

In [14]:
numbers = np.array([3, -1, 7, 8, 10, 11, 2, 7, 5])
scores = np.array([10, 8.5, 7.8, 9, 9.2, 6.7])
names = np.array(["Алёна", "Борис", "Василий", "Георгий", "Даниил"])
column = np.array([1, 2, 3, 4, 5, "id6"])

In [15]:
print(numbers, numbers.dtype)
print(scores, scores.dtype)
print(names, names.dtype)
print(column, column.dtype)

[ 3 -1  7  8 10 11  2  7  5] int64
[10.   8.5  7.8  9.   9.2  6.7] float64
['Алёна' 'Борис' 'Василий' 'Георгий' 'Даниил'] <U7
['1' '2' '3' '4' '5' 'id6'] <U21


> В Python существует иерархия типов. В случае, если в массиве встречаются элементы разных типов, более сильный тип вытесняет более слабый. Дробные числа (тип `float`) сильнее целых (тип `integer`), строковый тип (тип `str`) сильнее числовых. Поэтому в массиве `scores` все значения стали дробными (у изначально целочисленных элементов добавилась нулевая дробная часть и точка), а в массиве `column` все значения превратились в текстовые. Запись `<U7` означает, что все элементы являются последовательностями символов длины не больше 7 (самое длинное слово из 7 букв) в кодировке Unicode. С записью `<U21` аналогичная история, просто из-за того, что в массиве изначально вперемешку с текстом были числа, NumPy зарезервировал побольше места на всякий случай, под 21 символ.

### Задача 2. Money, money

В массиве `money` сохранена информация о стоимости покупок (в рублях), совершенных посетителями магазина. 
Создайте переменную `m` и сохраните в нее какое-нибудь число не менее 500.

* сохраните в массив `sdacha` суммы, которые посетители получат как сдачу с `m` рублей
* сохраните в массив `sotki` количество купюр номиналом 100 рублей, которые посетители могут получить как сдачу с `m` рублей.

In [16]:
money = np.array([125, 288, 324, 460, 128, 344, 45, 235, 336])
m = 500

In [17]:
sdacha = m - money
print(sdacha)

[375 212 176  40 372 156 455 265 164]


In [18]:
# оператор // – целочисленное деление

sotki = sdacha // 100
print(sotki)

[3 2 1 0 3 1 4 2 1]


### Задача 3. Мама, я политолог

В трех массивах сохранены статистические данные по одним и тем же избирательным участкам:

* `valid` – количество действительных бюллетеней;
* `invalid` – количество недействительных бюллетеней;
* `total` – общее количество зарегистрированных избирателей.

Получите массив `turnout` с явкой на этих избирательных участках (в голосах) и массив `turnout_perc` с явкой (в процентах).

In [19]:
valid = np.array([125, 132, 140, 164, 148, 190, 23, 140, 78, 20])
invalid = np.array([2, 5, 1, 4, 0, 10, 0, 3, 2, 0])
total = np.array([262, 310, 253, 440, 528, 367, 23, 314, 142, 22])

In [20]:
turnout = valid + invalid
turnout_perc = turnout / total * 100
print(turnout)
print(turnout_perc)

[127 137 141 168 148 200  23 143  80  20]
[ 48.47328244  44.19354839  55.7312253   38.18181818  28.03030303
  54.49591281 100.          45.54140127  56.33802817  90.90909091]


### Задача 4. Мама, я диетолог

В массивах `weight` и `height` сохранены значения веса (в килограммах) и роста (в сантиметрах) пациентов. Сохраните в массив `bmi` значения индекса массы тела для этих пациентов ($m$ – вес в килограммах, $h$ – рост в метрах):

$$
BMI = \frac{m}{h^2}
$$

In [21]:
weight = np.array([49, 62, 89, 72, 96, 48, 62, 55])
height = np.array([168, 178, 175, 180, 183, 157, 172, 164])
print(len(weight) == len(height))

True


In [22]:
bmi = weight / (height / 100) ** 2
print(bmi)

[17.36111111 19.56823633 29.06122449 22.22222222 28.66612918 19.47340663
 20.9572742  20.44913742]


Вычислите число и долю респондентов, чей индекс массы тела:

* ниже нормы (менее 18.5)
* выше нормы (25 и выше)

In [23]:
# шаг 1: изучаем выполнения условия
# условие проверяется для всех элементов,
# получается массив из True и False

bmi < 18.5

array([ True, False, False, False, False, False, False, False])

In [24]:
# шаг 2: подставляем условие в фильтр – квадратные скобки
# считаем число элементов – длину массива answer01
# считаем долю элементов – делим на общее число пациентов

answer01 = bmi[bmi < 18.5]
print(len(answer01))
print(len(answer01) / len(bmi))

1
0.125


In [25]:
# шаг 3: решаем задачу более эффективно

# массив из True и False для Python эквивалентен массиву из 1 (True) и 0 (False), 
# значит, для подсчета числа True (1) можно просто сложить все его элементы,
# 1 сложатся, а 0 на сумму не повлияют

print(bmi < 18.5)
print((bmi < 18.5).sum())

# доля True (1) получается делением числа True на общее количество элементов,
# число True (1) выше посчитали как сумму элементов массива из True и False,
# значит, доля True (1) равна среднему по массиву из True и False

print((bmi < 18.5).mean())

[ True False False False False False False False]
1
0.125


In [26]:
# то же для второго пункта

print(bmi >= 25)
print((bmi >= 25).sum())
print((bmi >= 25).mean())

[False False  True False  True False False False]
2
0.25


### Задача 5. Доброе утро

В задачнике по теории вероятностей когда-то была задача про Алису, которая встает утром то с правой, то с левой ноги. В массиве `experiment` зафиксированы результаты 30-дневного наблюдения за Алисой, Вычислите процент дней, когда Алиса вставала с правой ноги.

In [27]:
experiment = np.array(["right", "left", "right", "right", "left", 
                      "left", "left", "right", "left", "right", 
                      "right", "right", "right", "right", "left",
                      "left", "left", "right", "left", "right",
                      "right", "right", "right", "right", "right",
                      "left", "left", "right", "right", "right"])

In [28]:
# как вычислить долю быстро – в предыдущей задаче,
# здесь долю домножили на 100

p = (experiment == "right").mean() * 100
print(p)

63.33333333333333
